akka-persistence/src/main/scala/akka/persistence/journal/AsyncRecovery.scala

```diff
@@ -70,3 +70,47 @@ trait AsyncRecovery {
   def asyncReadHighestSequenceNr(persistenceId: String, fromSequenceNr: Long): Future[Long]
   //#journal-plugin-api
 }
+
+/**
+ * A plugin may implement this trait as an optimization. Combining `asyncReplayMessages` and
+ * `asyncReadHighestSequenceNr` into one method. If this trait is implemented the
+ * methods in [[AsyncRecovery]] will not be called.
+ */
+trait AsyncReplay {
+
+  /**
+   * Plugin API: asynchronously replays persistent messages. Implementations replay
+   * a message by calling `replayCallback`. The returned future must be completed
+   * with the highest sequence number when all messages (matching the sequence number bounds)
+   * have been replayed. Journal must maintain the highest sequence number and never decrease it.
+   * The future must be completed with a failure if any of the persistent messages
+   * could not be replayed.
+   *
+   * The `toSequenceNr` is the lowest of what was returned by [[#asyncReadHighestSequenceNr]]
+   * and what the user specified as recovery [[akka.persistence.Recovery]] parameter.
+   * This does imply that this call is always preceded by reading the highest sequence
+   * number for the given `persistenceId`.
```

Comment:
This part copied over from `asyncReplayMessages`?

Should it instead say that `toSequenceNr` will either be `Long.MaxValue` (so can return last replayed as highest) or otherwise it's been limited explicitly, in which case an extra database query is likely needed within this method to get the actual highest sequence number?

Reply:
Yes, that was obsolete. I clarified.